Dataset : Retail Sales Dataset

About : It is an Synthetic Dataset of 4310 rows x 21 columns containing the data regarding the orders.

## Environment Setup

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Load the dataset

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Sales Retail Dataset/retail_sales_dataset.csv')
display(df.head())

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,ORD-03903,2024-08-15,CUST32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,ORD-02198,2022-08-16,CUST07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,ORD-01348,2021-08-29,CUST49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,ORD-02076,2022-07-05,CUST05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,ORD-00287,2020-05-07,CUST58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


## Basic understanding of the dataset

display the top 5 rows of the dataset

In [ ]:
df.head()

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,ORD-03903,2024-08-15,CUST32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,ORD-02198,2022-08-16,CUST07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,ORD-01348,2021-08-29,CUST49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,ORD-02076,2022-07-05,CUST05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,ORD-00287,2020-05-07,CUST58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


Checking number of rows and columns present in the dataset

In [ ]:
df.shape

(4310, 21)

Overview of the dataset

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4310 entries, 0 to 4309
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               4280 non-null   object 
 1   order_date             4280 non-null   object 
 2   customer_id            4280 non-null   object 
 3   customer_name          4280 non-null   object 
 4   age                    4150 non-null   float64
 5   gender                 4280 non-null   object 
 6   region                 4280 non-null   object 
 7   city                   4280 non-null   object 
 8   product_category       4280 non-null   object 
 9   product_name           4280 non-null   object 
 10  quantity               4170 non-null   float64
 11  unit_price             4280 non-null   float64
 12  discount_pct           4143 non-null   float64
 13  sales_amount           4280 non-null   float64
 14  profit                 4280 non-null   float64
 15  ship

1. RangeIndex: 4310 Rows, 0 to 4309
2. Total columns 21
3. Here all 21 columns contains Null values.
4. Here there are two data types : a. float64 (9) and b. object (12)
5. Memory Usage : 707.2+ KB

Checking for null values based on each column

In [ ]:
df.isnull().sum()

,0
order_id,30
order_date,30
customer_id,30
customer_name,30
age,160
gender,30
region,30
city,30
product_category,30
product_name,30


Find duplicate rows

In [ ]:
df.duplicated().sum()

np.int64(109)

Here in the dataset we have 109 duplicate values

Descriptive statitics of the dataset(Only numerical columns)

In [ ]:
df.describe()

,age,quantity,unit_price,discount_pct,sales_amount,profit,shipping_cost,customer_satisfaction,days_to_ship
count,4150.000000,4170.000000,4280.000000,4143.000000,4.280000e+03,4.280000e+03,4280.000000,3932.000000,4180.000000
mean,35.335181,7.379856,10621.154428,0.202085,6.235864e+04,1.168713e+04,63.358836,3.016531,5.620335
std,35.709382,43.578352,17323.704016,0.118095,3.166998e+05,5.287522e+04,18.312281,1.413127,4.178663
min,-7.000000,-1.000000,50.260000,0.000000,3.945000e+01,1.560000e+00,-4.290000,1.000000,-1.000000
25%,26.000000,3.000000,672.222500,0.100000,2.462165e+03,7.018225e+02,50.817500,2.000000,3.000000
50%,34.000000,5.000000,2228.435000,0.210000,9.548075e+03,3.128005e+03,63.465000,3.000000,6.000000
75%,42.000000,8.000000,12717.435000,0.310000,5.006964e+04,1.040372e+04,75.650000,4.000000,8.000000
max,999.000000,999.000000,79957.920000,0.400000,1.736031e+07,2.445648e+06,133.350000,5.000000,100.000000


1. Count : Number of non-null values
2. mean : Avergae value
3. std : Standard deviation (spread of data)
4. min : Smallest Value
5. 25% : First Quartile (Q1)
6. 50% : Median (Q2)
7. 75% : Third Quartile (Q3)
8. max : Largest Value

# Cleaning Notes



1. order_id: 30 missing values - Remove ORD and convert dtype and drop duplicates.
2. order_date: 30 missing values - Date format issue
3. customer_id: 30 missing values - Remove CUST and convert dtype.
4. customer_name: 30 missing values - Ensure proper case.
5. age: 160 Missing value. Convert data type from Float to Int
6. gender: Convert into Int by changing them into 1,2,3
7. region: 30 Missing Value & Convert Central by C, and so on
8. city : 30 NAN, Check city is in Proper Case
9. product_category : 30 Nan
10. product_name : 30 Null value
11. quantity : 140 NAN, Few are in -ve, Convert into Int
12. unit_price : 30 NAN, All time should be > 0
13. discount_pct: 167 NAN, All time should be > 0 and <100
14. sales_amount 30 NAN, All time should be > 0
15. profit 30 NAN,
16. shipping_cost 30, NAN All time should be > 0
17. payment_method : 30 NAN
19. customer_satisfaction : 378 NAN Float to Int
20. return_flag 30 NAN
21. order_status 30 NAN, Case issues
22. days_to_ship 130 NAN; All time should be > 0

# 1. Data Cleaning: order_id

Step 1: remove "ORD-" from all the values in Order_id column

In [ ]:
try:
  df['order_id'] = df['order_id'].str.replace("ORD-0", "", regex = False)
  display(df.head())
except:
  print("Already Removed")

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,3903,2024-08-15,CUST32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,2198,2022-08-16,CUST07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,1348,2021-08-29,CUST49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,2076,2022-07-05,CUST05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,0287,2020-05-07,CUST58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


Step 2: Drop rows where dupliucate order_id present

In [ ]:
df = df.drop_duplicates(subset = 'order_id')
display(df.head())

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,3903,2024-08-15,CUST32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,2198,2022-08-16,CUST07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,1348,2021-08-29,CUST49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,2076,2022-07-05,CUST05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,0287,2020-05-07,CUST58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


In [ ]:

df['order_id'].describe()

,order_id
count,4200
unique,4200
top,3241
freq,1


Step 3: Remove NAN

In [ ]:
df['order_id'].isnull().sum()

np.int64(1)

In [ ]:
df[df['order_id'].isna()]

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
66,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
df = df.dropna(subset = ['order_id'])
df[df['order_id'].isna()]

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship


Step 4: Change the data type

In [ ]:
df['order_id'].dtypes

dtype('O')

Convert Data type from Object to Int64

In [ ]:
df['order_id'] = df['order_id'].astype(int)
df['order_id'].dtypes

dtype('int64')

# 2. Data Cleaning - order_date

In [ ]:
df['order_date'].unique()

array(['2024-08-15', '2022-08-16', '2021-08-29', ..., '2020-12-16',
       '16/09/2023', '2022-02-11'], dtype=object)

Convert all order_date in 'YYYY-MM-DD' format

In [ ]:
df['order_date'] = pd.to_datetime(df['order_date'], errors= 'coerce').dt.strftime('%Y-%m-%d')
df['order_date'].dtypes

dtype('O')

Convert all order_date in date format

In [ ]:
df['order_date'] = pd.to_datetime(df['order_date'])
df['order_date'].dtypes

dtype('<M8[ns]')

Count the NAN values in order_date column

In [ ]:
df['order_date'].isnull().sum()

np.int64(267)

Check the frequency of 'order_date'

In [ ]:
df['order_date'].value_counts().sort_values()

,count
order_date,
2023-04-08,1
2024-01-19,1
2023-04-27,1
2020-11-21,1
2020-07-04,1
...,...
2023-01-27,8
2022-11-28,8
2024-08-09,8


Replace Nan value of order_date by '2023-01-12'(Mode)

In [ ]:
df['order_date'].fillna('2023-01-12', inplace = True)

/tmp/ipykernel_1153/1785927064.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['order_date'].fillna('2023-01-12', inplace = True)


In [ ]:
df['order_date'].isnull().sum()

np.int64(0)

In [ ]:
df.isnull().sum()

,0
order_id,0
order_date,0
customer_id,0
customer_name,0
age,125
gender,0
region,0
city,0
product_category,0
product_name,0


# 3. Data Cleaning - customer_id

Step 1: Remove "CUST" from all the values in customer_id column

In [ ]:
try:
  df['customer_id'] = df['customer_id'].str.replace("CUST", "", regex = False)
  display(df.head())
except:
  print("Already Removed")

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,3903,2024-08-15,32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,2198,2022-08-16,07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,1348,2021-08-29,49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,2076,2022-07-05,05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,287,2020-05-07,58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


Step 2: Count and remove the NAN values

In [ ]:
df['customer_id'].isnull().sum()

np.int64(0)

In [ ]:
df = df.dropna(subset = ['customer_id'])
df['customer_id'].isnull().sum()

np.int64(0)

Step 3: Convert data type from Object to Int64

In [ ]:
df['customer_id'] = df['customer_id'].astype(int)
df['customer_id'].dtypes

dtype('int64')

# 4. Data Cleaning - customer_name

Step 1: Count the NAN values

In [ ]:
df['customer_name'].isnull().sum()

np.int64(0)

Step 2: Replace NAN with 'Unknown' and ensure Proper Case

In [ ]:
df['customer_name'] = df['customer_name'].fillna("Unknown")
df['customer_name'] = df['customer_name'].str.strip().str.title()
df['customer_name'].unique()[:10]

array(['Aarav Mehta', 'Sanjay Verma', 'Pooja Sharma', 'Priya Reddy',
       'Ritu Sharma', 'Vikram Iyer', 'Arjun Shah', 'Aditya Sharma',
       'Aditya Patel', 'Arjun Das'], dtype=object)

# 5. Data Cleaning - age

Step 1: Check the value range (we found invalid values like -7 and 999)

In [ ]:
print('Min:', df['age'].min(), 'Max:', df['age'].max())
df['age'].isnull().sum()

Min: -7.0 Max: 999.0


np.int64(125)

Step 2: Replace invalid ages (age < 0 or age > 100) with NAN

In [ ]:
df.loc[(df['age'] < 0) | (df['age'] > 100), 'age'] = np.nan

Step 3: Fill NAN with the median age

In [ ]:
median_age = df['age'].median()
df['age'] = df['age'].fillna(median_age)
print('Median used:', median_age)
df['age'].isnull().sum()

Median used: 34.0


np.int64(0)

Step 4: Convert data type from Float to Int

In [ ]:
df['age'] = df['age'].astype(int)
df['age'].dtypes

dtype('int64')

# 6. Data Cleaning - gender

In [ ]:
df['gender'].unique()

array(['Other', 'Male', 'Female', 'FEMALE', 'f', 'F', 'male', 'female',
       'MALE', 'M', 'm'], dtype=object)

Step 1: Standardise text to a single clean form (Male / Female / Other)

In [ ]:
df['gender'] = df['gender'].str.strip().str.upper()

gender_map = {
    'M': 'Male', 'MALE': 'Male',
    'F': 'Female', 'FEMALE': 'Female',
    'OTHER': 'Other'
}
df['gender'] = df['gender'].map(gender_map)
df['gender'].unique()

array(['Other', 'Male', 'Female'], dtype=object)

Step 2: Fill NAN with 'Other'

In [ ]:
df['gender'] = df['gender'].fillna("Other")
df['gender'].isnull().sum()

np.int64(0)

Step 3: Convert into Int by mapping Male -> 1, Female -> 2, Other -> 3

In [ ]:
gender_code = {'Male': 1, 'Female': 2, 'Other': 3}
df['gender'] = df['gender'].map(gender_code).astype(int)
df['gender'].unique()

array([3, 1, 2])

7. Data Cleaning - region

Step 1: Count the NAN values

In [ ]:
df['region'].isnull().sum()

np.int64(0)

Step 2: Fill NAN with 'Unknown' and standardise case

In [ ]:
df['region'] = df['region'].fillna("Unknown").str.strip().str.title()
df['region'].unique()

array(['Central', 'West', 'North', 'East', 'South'], dtype=object)

Step 3: Convert region into short codes (Central -> C, West -> W, North -> N, East -> E, South -> S)

In [ ]:
region_map = {
    'Central': 'C', 'West': 'W', 'North': 'N',
    'East': 'E', 'South': 'S', 'Unknown': 'U'
}
df['region'] = df['region'].map(region_map)
df['region'].unique()

array(['C', 'W', 'N', 'E', 'S'], dtype=object)

# 8. Data Cleaning - city

Step 1: Count the NAN values

In [ ]:
df['city'].isnull().sum()

np.int64(0)

Step 2: Fill NAN with 'Unknown' and ensure Proper Case

In [ ]:
df['city'] = df['city'].fillna("Unknown").str.strip().str.title()
df['city'].unique()[:15]

array(['Raipur', 'Mumbai', 'Surat', 'Ahmedabad', 'Nagpur', 'Bhopal',
       'Amritsar', 'Patna', 'Ludhiana', 'Pune', 'Kochi', 'Bangalore',
       'Guwahati', 'Delhi', 'Chennai'], dtype=object)

# 9. Data Cleaning - product_category

Step 1: Count the NAN values

In [ ]:
df['product_category'].isnull().sum()

np.int64(0)

Step 2: Fill NAN with 'Unknown' and ensure Proper Case

In [ ]:
df['product_category'] = df['product_category'].fillna("Unknown").str.strip().str.title()
df['product_category'].unique()

array(['Groceries', 'Beauty', 'Furniture', 'Books', 'Clothing',
       'Electronics', 'Sports'], dtype=object)

# 10. Data Cleaning - product_name

Step 1: Count the NAN values

In [ ]:
df['product_name'].isnull().sum()

np.int64(0)

Step 2: Fill NAN with 'Unknown' and ensure Proper Case

In [ ]:
df['product_name'] = df['product_name'].fillna("Unknown").str.strip().str.title()
df['product_name'].isnull().sum()

np.int64(0)

# 11. Data Cleaning - quantity

Step 1: Check the range (some values are negative)

In [ ]:
print('Min:', df['quantity'].min(), 'Max:', df['quantity'].max())
df['quantity'].isnull().sum()

Min: -1.0 Max: 999.0


np.int64(108)

Step 2: Replace invalid quantity (<= 0) with NAN

In [ ]:
df.loc[df['quantity'] <= 0, 'quantity'] = np.nan

Step 3: Fill NAN with the median quantity

In [ ]:
median_qty = df['quantity'].median()
df['quantity'] = df['quantity'].fillna(median_qty)
print('Median used:', median_qty)
df['quantity'].isnull().sum()

Median used: 5.0


np.int64(0)

Step 4: Convert data type from Float to Int

In [ ]:
df['quantity'] = df['quantity'].astype(int)
df['quantity'].dtypes

dtype('int64')

# 12. Data Cleaning - unit_price

Step 1: Count NAN and check all values are > 0

In [ ]:
print('NAN:', df['unit_price'].isnull().sum())
print('Values <= 0:', (df['unit_price'] <= 0).sum())

NAN: 0
Values <= 0: 0


Step 2: Replace any value <= 0 with NAN, then fill all NAN with the median

In [ ]:
df.loc[df['unit_price'] <= 0, 'unit_price'] = np.nan
median_price = df['unit_price'].median()
df['unit_price'] = df['unit_price'].fillna(median_price)
df['unit_price'].isnull().sum()

np.int64(0)

# 13. Data Cleaning - discount_pct

Note: discount is stored as a fraction (0 to 1), so valid range is >= 0 and <= 1.

Step 1: Count NAN and check the range

In [ ]:
print('NAN:', df['discount_pct'].isnull().sum())
print('Min:', df['discount_pct'].min(), 'Max:', df['discount_pct'].max())

NAN: 135
Min: 0.0 Max: 0.4


Step 2: Replace out-of-range values (< 0 or > 1) with NAN, then fill NAN with 0 (no discount)

In [ ]:
df.loc[(df['discount_pct'] < 0) | (df['discount_pct'] > 1), 'discount_pct'] = np.nan
df['discount_pct'] = df['discount_pct'].fillna(0)
df['discount_pct'].isnull().sum()

np.int64(0)

# 14. Data Cleaning - sales_amount

Step 1: Count NAN and check all values are > 0

In [ ]:
print('NAN:', df['sales_amount'].isnull().sum())
print('Values <= 0:', (df['sales_amount'] <= 0).sum())

NAN: 0
Values <= 0: 0


Step 2: Replace any value <= 0 with NAN, then fill all NAN with the median

In [ ]:
df.loc[df['sales_amount'] <= 0, 'sales_amount'] = np.nan
df['sales_amount'] = df['sales_amount'].fillna(df['sales_amount'].median())
df['sales_amount'].isnull().sum()

np.int64(0)

# 15. Data Cleaning - profit

Note: profit can legitimately be negative (a loss), so we only handle NAN values here.

Step 1: Count the NAN values

In [ ]:
df['profit'].isnull().sum()

np.int64(0)

Step 2: Fill NAN with the median profit

In [ ]:
df['profit'] = df['profit'].fillna(df['profit'].median())
df['profit'].isnull().sum()

np.int64(0)

# 16. Data Cleaning - shipping_cost

Step 1: Count NAN and check all values are > 0

In [ ]:
print('NAN:', df['shipping_cost'].isnull().sum())
print('Values <= 0:', (df['shipping_cost'] <= 0).sum())

NAN: 0
Values <= 0: 2


Step 2: Replace any value <= 0 with NAN, then fill all NAN with the median

In [ ]:
df.loc[df['shipping_cost'] <= 0, 'shipping_cost'] = np.nan
df['shipping_cost'] = df['shipping_cost'].fillna(df['shipping_cost'].median())
df['shipping_cost'].isnull().sum()

np.int64(0)

#  17. Data Cleaning - payment_method

Step 1: Count the NAN values

In [ ]:
df['payment_method'].isnull().sum()

np.int64(0)

Step 2: Fill NAN with 'Unknown' and ensure Proper Case

In [ ]:
df['payment_method'] = df['payment_method'].fillna("Unknown").str.strip().str.title()
df['payment_method'].unique()

array(['Debit Card', 'Emi', 'Net Banking', 'Cash On Delivery',
       'Credit Card', 'Upi'], dtype=object)

# 18. Data Cleaning - customer_satisfaction

Step 1: Count the NAN values (rating scale is 1 to 5)

In [ ]:
df['customer_satisfaction'].isnull().sum()

np.int64(346)

Step 2: Fill NAN with the median rating

In [ ]:
median_sat = df['customer_satisfaction'].median()
df['customer_satisfaction'] = df['customer_satisfaction'].fillna(median_sat)
print('Median used:', median_sat)
df['customer_satisfaction'].isnull().sum()

Median used: 3.0


np.int64(0)

Step 3: Convert data type from Float to Int

In [ ]:
df['customer_satisfaction'] = df['customer_satisfaction'].astype(int)
df['customer_satisfaction'].dtypes

dtype('int64')

# 19. Data Cleaning - return_flag

Step 1: Count the NAN values

In [ ]:
df['return_flag'].isnull().sum()

np.int64(0)

Step 2: Fill NAN with False (assume not returned) and convert to int (0/1)

In [ ]:
df['return_flag'] = df['return_flag'].fillna(False).astype(int)
df['return_flag'].unique()

/tmp/ipykernel_1153/3652210849.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['return_flag'] = df['return_flag'].fillna(False).astype(int)


array([1, 0])

# 20. Data Cleaning - order_status

Check unique values - they have case issues (delivered, DELIVERED, etc.)

In [ ]:
df['order_status'].unique()

array(['Returned', 'Delivered', 'Pending', 'Cancelled', 'Shipped',
       'delivered', 'cancelled', 'shipped', 'pending'], dtype=object)

Step 1: Fill NAN with 'Unknown' and fix the case issues using Proper Case

In [ ]:
df['order_status'] = df['order_status'].fillna("Unknown").str.strip().str.title()
df['order_status'].unique()

array(['Returned', 'Delivered', 'Pending', 'Cancelled', 'Shipped'],
      dtype=object)

# 21. Data Cleaning - days_to_ship

Step 1: Count NAN and check all values are > 0

In [ ]:
print('NAN:', df['days_to_ship'].isnull().sum())
print('Values <= 0:', (df['days_to_ship'] <= 0).sum())

NAN: 97
Values <= 0: 7


Step 2: Replace any value <= 0 with NAN, then fill all NAN with the median

In [ ]:
df.loc[df['days_to_ship'] <= 0, 'days_to_ship'] = np.nan
median_days = df['days_to_ship'].median()
df['days_to_ship'] = df['days_to_ship'].fillna(median_days)
print('Median used:', median_days)
df['days_to_ship'].isnull().sum()

Median used: 6.0


np.int64(0)

Step 3: Convert data type from Float to Int

In [ ]:
df['days_to_ship'] = df['days_to_ship'].astype(int)
df['days_to_ship'].dtypes

dtype('int64')

# Final Verification - Cleaned Dataset

Check that there are no NAN values remaining in any column

In [ ]:
df.isnull().sum()

,0
order_id,0
order_date,0
customer_id,0
customer_name,0
age,0
gender,0
region,0
city,0
product_category,0
product_name,0


Check the final data types of all columns

In [ ]:
df.dtypes

,0
order_id,int64
order_date,datetime64[ns]
customer_id,int64
customer_name,object
age,int64
gender,int64
region,object
city,object
product_category,object
product_name,object


Final shape and a preview of the cleaned dataset

In [ ]:
print('Final shape:', df.shape)
df.head()

Final shape: (4200, 21)


,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,3903,2024-08-15,32603,Aarav Mehta,39,3,C,Raipur,Groceries,Sugar,...,346.32,0.00,4043.14,289.18,56.04,Debit Card,3,1,Returned,7
1,2198,2022-08-16,7881,Sanjay Verma,28,1,W,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5,0,Delivered,2
2,1348,2021-08-29,49296,Pooja Sharma,41,3,W,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4,0,Delivered,5
3,2076,2022-07-05,5587,Priya Reddy,29,3,W,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,Emi,3,0,Delivered,4
4,287,2020-05-07,58035,Ritu Sharma,9,1,C,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4,0,Pending,4


Save the cleaned dataset to a new CSV file

In [ ]:
import os

output_dir = '/content/drive/MyDrive/Sales Retail Dataset'
os.makedirs(output_dir, exist_ok=True)

output_path = os.path.join(output_dir, 'retail_sales_cleaned.csv')
df.to_csv(output_path, index = False)
print('Cleaned dataset saved successfully!')

Cleaned dataset saved successfully!
